### PR TITLE
[AutoTVM] Support to pass additional compiler options

### PR DESCRIPTION
**Pull request by FrozenGene**
Currently, AutoTVM doesn't support to pass additional compiler options to builder. It works fine on most cases. However, if we use NDK CC, some times we want to specify additional compiler options. For example, another `sysroot`.

After this PR, the user could do like this:

```python
os.environ['TVM_NDK_CC'] = /path/to/ndk_cc
options=['--sysroot=/path/to/sysroot', '-shared', '-fPIC']

tuning_option = {
    // ...
    'measure_option': autotvm.measure_option(
        builder=autotvm.LocalBuilder(
            build_func='ndk' if use_android else 'default', options=options),
       ...
    ),
}
```
@merrymercy @tqchen @ajtulloch would you like to help review it? Thanks.

### CODE REVIEW

File: python/tvm/module.py [hunk 1]
```diff
@@ -162,13 +162,20 @@ def export_library(self,
...
+        if "options" in kwargs or has_c_module:
+            opts = kwargs["options"] if "options" in kwargs else None
+            options = opts if isinstance(opts, (list, tuple)) else [opts]
+            options = [opt for opt in options if opt]
+            if has_c_module:
+                options = options + ["-I" + path for path in find_include_path()]
+            if not options:
+                del kwargs["options"]
+            else:
+                kwargs.update({'options': options})
```
Comment (comaniac) [inline python/tvm/module.py]: I feel like we don't have to test `has_c_module` twice. Maybe the following logic is sufficient?
```
options = ["-I" +path for path in find_include_path()] if has_c_module else []
if "options" in kwargs:
  opts = kwargs["options"]
  options += opts if isinstance(opts, (list, tuple)) else [opts]
  options = [opt for opt in options if opt]
  if not options:
    del kwargs["options"]
if options:
  kwargs.update({"options": options})
```

Comment (FrozenGene) [reply python/tvm/module.py]: Yes. Your way could avoid twice check of `has_c_module`.

Comment (FrozenGene) [reply python/tvm/module.py]: Done. Please review it again.

**Comment by merrymercy** [comment]
1. We should not put `options` in `MeasureResult`. It is not result (output). It is input.
2. If you change MeasureInput / MeasureResult, you should also update serialization for them in `python/tvm/autotvm/record.py::encode, decode`. You should be very careful about the backward-compatibility.
3. If you add `options` to records, why don't you also add `build_func`?

Currently, I think not serializing them is fine. Or we can embed `options` into `MeasureInput.task` and serialize them carefully.

**Comment by tqchen** [comment]
Likely you can create a build function(via functor that captures these additional options) instead of directly passing ndk

**Comment by tqchen** [comment]
Try this:

```python
from tvm.contrib import cc, ndk

build_func = cc.cross_compiler(
   ndk.create_shared, options)
```

**Comment by FrozenGene** [comment]
> Try something along the following line:
> 
> ```python
> from tvm.contrib import cc, ndk
> 
> build_func = cc.cross_compiler(ndk.create_shared, options)
> 
> tuning_option = {
>     // ...
>     'measure_option': autotvm.measure_option(
>         builder=autotvm.LocalBuilder(
>             build_func=build_func if use_android else 'default', options=options),
>        ...
>     ),
> }
> ```
> 
> Perhaps it is a good time to document these usecases of cross_compiler(e.g. get target triple from compile_func if it has the attr)

Yes. Your way is good and don't change anything. Do you mean this ? 
```python
from tvm.contrib import cc, ndk

build_func = cc.cross_compiler(ndk.create_shared, options)

tuning_option = {
    // ...
    'measure_option': autotvm.measure_option(
        builder=autotvm.LocalBuilder(
            build_func=build_func if use_android else 'default'),
       ...
    ),
}
```
No options needed in the builder.

**Comment by FrozenGene** [comment]
> 1. We should not put `options` in `MeasureResult`. It is not the result (output). It is input.
> 2. If you change MeasureInput / MeasureResult, you should also update serialization for them in `python/tvm/autotvm/record.py::encode, decode`. You should be very careful about the backward-compatibility.
> 3. If you add `options` to records, why don't you also add `build_func`?
> 
> Currently, I think not serializing them is fine. Or we can attach `options` into `MeasureInput.task` and serialize them carefully.

Yeah. I agree with you. I think this PR is not one elegant solution for this problem. The nicer way is @tqchen 's suggestion, using custom build func. I will close this PR. Thanks.

**Pull request closed without merging**